### ISSUE TITLE
Support for enum in OpenAPI

### ISSUE THREAD

**Issue by guersam** [julienrf/endpoints#133]
As discussed in [gitter](https://gitter.im/julienrf/endpoints?at=5bbae6cebbdc0b25050246fb), it'll be nice to add [enum](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#fixed-fields-5) to the core algebra. 

Things to consider:

1. Do we need `type Enum[A] <: JsonSchema[A]`, or is `def enum(...): Tagged[A]` just enough?
2. Do we integrate it with scala Enumeration?
3. Do we integrate it with 3rd party enum lib(s)?
   - https://github.com/lloydmeta/enumeratum
   - https://github.com/scalalandio/enumz

**Comment by stefanholzmueller** [comment]
1. I don't see how `Tagged` helps here, since there is no "tag", just a set of possible values. From my  experience, I mostly use sealed traits with case objects to model enums, and map the objects to strings in the schema. `type Enum[A] <: JsonSchema[A]` sounds like a good approach to me, and I'm willing to participate.

2. I'd vote no, because it is kind of an unpopular part of Scala IMHO
3. Also no, but if users want to use these libraries, it should be possible

**Comment by julienrf** [comment]
I think it would make sense to model enumerations with the existing `Tagged[A]` type, for the sake of not introducing additional concepts. However, if we look at the current algebra it has the following operations involving `Tagged`:

~~~ scala
  def withDiscriminator[A](tagged: Tagged[A], discriminatorName: String): Tagged[A]
  def choiceTagged[A, B](taggedA: Tagged[A], taggedB: Tagged[B]): Tagged[Either[A, B]]
~~~

I’m not sure these operations make sense in the case of enums. I would just start with introducing a new `Enum[A]` type and we could always see if there is room for generalization afterward.

**Comment by stefanholzmueller** [comment]
I played around with this idea. Since I believe that sealed traits are the most general design, this is my test enum:
```scala
sealed trait Status
case object Status1 extends Status
case object Status2 extends Status
case object Status3 extends Status
object Status {
  def fromString(s: String): Status = s match {
    case "Status1" => Status1
    case "Status2" => Status2
    case "Status3" => Status3
  }
}
```
After a while I came up with this API:
```scala
enum[String]("status", List(Status1.toString, Status2.toString), Some("enum doc"))
  .invmap[Status](Status.fromString)(_.toString)
```
I liked this approach because it reuses the concept of `invmap` and it allows to map the enum to any primivite type (string in this case, but could also be a number). It is also possible to allow only a subset of the enum values, but unfortunately with this design the list of possible values have to be strings. Also, the number of different `invmap` and `Ops` is getting confusing.

An alternative API might look like this:
```scala
enum[Status, String]("status", List(Status1, Status2), Status.fromString, _.toString, Some("enum doc"))
```
So no `invmap` anymore, all in one (arguably too large) method.

A non-obvious difficulty is the interaction of `type Enum[A] <: JsonSchema[A]` and `type Record[A] <: JsonSchema[A]`. I cannot easily use `zip` or `:*:` because those only work on `Record`, and if I declare `Enum` as a subtype of `Record`, I get some problems with case-to-case inheritance in the openapi interpreter. It's a bit of a mess TBH.

My gut feeling is that `zip` should work on the common type `JsonSchema`, but there are probably good reasons for `type Record[A] <: JsonSchema[A]`. I've asked this before but forget the exact reason - something with openapi I think. And I'm shying away from a complete revamp of the `JsonSchemas` trait...

**Comment by julienrf** [comment]
I have a few comments:

1. The `fromString` function should be total: it should return an `Option[Status]`
2. Consequently, the approach that uses `invmap` can not be used
    - Note that we could simply introduce a new operation, similar to `invmap`, but that would support the transformation operation applied to the decoded value to fail again. We could name that operation `andThen`

**Comment by stefanholzmueller** [comment]
1. You're right, I agree that the "decode" function should return `Option`.
2. Can you give the type signature for `andThen`. I fail to see how it should be used.
3. Example:
```scala
case class Counter(value: Int, status: Status)

implicit lazy val jsonSchemaCounter: JsonSchema[Counter] = (
    field[Int]("value") zip
    enum[Status, String]("status", List(Status1, Status2), decode, encode, Some("enum doc"))
).invmap[Counter](Counter.tupled)(Counter.unapply(_).get)
```
For me, `enum` returns a field that can be `zip`ped together with other fields. `zip` expects two `Record` operands, so either the return value of `enum` needs to be `Record`, or we change `zip` to operate on `JsonSchema` (if that is even possible).

Hm, different idea: Maybe we can reuse `field` and `optField` (enums could be optional) somehow. So enums are like normal fields, except their possible values are constrained.

**Comment by julienrf** [comment]
2. Let’s name it `refined` instead of `andThen`:

~~~ scala
implicit class JsonSchemaOps[A](`this`: JsonSchema[A]) {

  def refined[B](f: A => Option[B], g: B => A): JsonSchema[B]

}
~~~

3. I think you want to write this instead:

~~~ scala
val jsonSchemaCounter = (
  field[Int]("value") :*:
  field[Status]("status", enum(…))
).as[Counter]
~~~

A JSON schema zipping a record and an enum would mean that the following would be a valid JSON document, but it’s not:

~~~ javascript
{
  "value": 42,
  "Status2"
}
~~~

Finally, with the `refine` method defined above we should be able to define enumerations as a refinement of `String` fields (so, no `Enum[A]` type would be required):

~~~ scala
implicit val jsonSchemaCounter: JsonSchema[Counter] = (
  field[Int]("value") :*:
  field[String]("status").refined(Status.fromString, _.toString)
).as[Counter]
~~~

(assuming that `fromString`) returns an `Option`.

WDYT? I’m not sure we can generate a meaningful documentation from the `refined` method.

**Comment by julienrf** [comment]
Yes, I think this is the way to go!

**Comment by julienrf** [comment]
Also note that the fromEnum and toEnum parameters are equivalent to an endpoints.Codec parameter.

**Comment by stefanholzmueller** [comment]
Turns out I could not use `Codec` easily, because adding `.dependsOnLocalCrossProjects("algebra")` in  subproject "json-schema" would cause a cyclic dependency. I'm now using `decode: A => Either[String, E]` (that function is supported by circe) and `encode: E => A`.

Supporting other types than strings is more difficult than I thought. Basically I'm carrying `values: Set[A]` around in the OpenAPI interpreter, and the compiler cannot find an implicit circe encoder for any `A`, obviously. I'm not happy with all the ideas I had so far. I'm tempted to give up and only support strings, which would simplify things. Personally I don't have a need for non-string enums. What do you think? @guersam @julienrf 

Another small question: I'm now using `Set` for the possible values, and then sort by the string representation. Would you prefer `List` instead, so that the user can easily define a certain order in which the values should appear in the OpenAPI documentation?

**Comment by julienrf** [comment]
> Basically I'm carrying values: Set[A] around in the OpenAPI interpreter, and the compiler cannot find an implicit circe encoder for any A

The strategy is to delay the requirement for the implicit `JsonSchema[A]` until `A` is effectively known, which should be the case as soon as you call `enum`, right?

Anyway, it’s more than fine to support only strings in a first version!

I don’t have a preference for `Set` over `Seq`. Pick whatever you think is best.

**Comment by guersam** [comment]
Agreed. 1) I either have string enums only and 2) `Seq` looks slightly better to me because many libraries expose the values in a Seq (or its subtype,) but `Set` is fine too because the users can specify the order via `ListSet` anyway.

**Comment by stefanholzmueller** [comment]
Thanks for your feedback!

@julienrf Yes, I tried to pass the `JsonSchema[A]` along, but ran into troubles. For future reference:

- The openapi interpreter uses a sealed trait and case classes without type parameters, I did not really want to add one.
- The openapi model does not use `JsonSchema` (it was not even available for import), but circe directly.

So I implemented only string enums now. Also, I decided to use `Seq` for the set of possible values.

**Comment by stefanholzmueller** [comment]
I just realized the `decode` function could be removed completely! Decoding will work by iterating over the possible values and calling the `encode` function on them, and check if it matches the raw string. Makes sense?

**Comment by guersam** [comment]
Do you want the decoding to be case sensitive, insensitive or customizable? For example, enumeratum explicitly provides both way via `.withName` and `.withNameInsensitive`. 

Our current implementation is case sensitive only but basically I think it's fine because many users will generate their client/server through the OpenAPI spec which guarantees the case will always match.

**Comment by stefanholzmueller** [comment]
I wanted the decoding to be always case sensitive, for exactly the reason you mentioned: OpenAPI enums are case-sensitive. The spec does not support documenting case-insensitive enums.

And I think it makes sense. It is simple and clear this way.
